analysis/tests/parameter_tuning_test.py

```diff
@@ -75,12 +76,114 @@ def test_find_candidate_parameters(
+            max_contributions_per_partition=True)
+
+        candidates = parameter_tuning._find_candidate_parameters(
+            mock_histograms,
+            parameters_to_tune,
+            pipeline_dp.Metrics.COUNT,
+            ParametersSearchStrategy.QUANTILES,
+            max_candidates=5)
+        self.assertEqual([1, 1, 2, 2], candidates.max_partitions_contributed)
+        self.assertEqual([4, 5, 4, 5],
+                         candidates.max_contributions_per_partition)
+
+    def test_find_candidate_parameters_constant_relative_step_strategy_big_n_max(
+            self):
+        mock_l0_histogram = histograms.Histogram(None, None)
+        setattr(histograms.Histogram, 'max_value', 999999)
+
+        mock_histograms = histograms.DatasetHistograms(mock_l0_histogram, None,
+                                                       None, None, None)
+        parameters_to_tune = parameter_tuning.ParametersToTune(
+            max_partitions_contributed=True,
+            max_contributions_per_partition=False)
+
+        candidates = parameter_tuning._find_candidate_parameters(
+            mock_histograms,
+            parameters_to_tune,
+            pipeline_dp.Metrics.COUNT,
+            ParametersSearchStrategy.CONSTANT_RELATIVE_STEP,
+            max_candidates=1000)
+
+        expected_superset = set(
+            list(range(1, 1000, 1)) + list(range(1000, 10000, 10)) +
+            list(range(10000, 100000, 100)) +
+            list(range(100000, 1000000, 1000))).union({999999})
+        self.assertTrue(
+            set(candidates.max_partitions_contributed).issubset(
+                expected_superset))
+        self.assertEqual(len(set(candidates.max_partitions_contributed)),
+                         len(candidates.max_partitions_contributed))
+        self.assertEqual(len(candidates.max_partitions_contributed), 1000)
+        self.assertEqual(sorted(candidates.max_partitions_contributed),
+                         candidates.max_partitions_contributed)
+
+    def test_find_candidate_parameters_constant_relative_step_strategy_small_n_max(
+            self):
+        mock_linf_histogram = histograms.Histogram(None, None)
+        setattr(histograms.Histogram, 'max_value', 999999)
+
+        mock_histograms = histograms.DatasetHistograms(None, None,
+                                                       mock_linf_histogram,
+                                                       None, None)
+        parameters_to_tune = parameter_tuning.ParametersToTune(
+            max_partitions_contributed=False,
+            max_contributions_per_partition=True)
+
+        candidates = parameter_tuning._find_candidate_parameters(
+            mock_histograms,
+            parameters_to_tune,
+            pipeline_dp.Metrics.COUNT,
+            ParametersSearchStrategy.CONSTANT_RELATIVE_STEP,
+            max_candidates=10)
+
+        self.assertEqual(10, len(candidates.max_contributions_per_partition))
```

Comment:
Nit: Please hard code the expected output, here there are only 10 elements, and it will be seen from the output how the algorithm generates them.

Reply:
I updated but actually I don't think it is easy to understand why elements are exactly these and not the others.